python/ray/data/dataset.py

```diff
@@ -3976,6 +3977,32 @@ def materialize(self) -> "MaterializedDataset":
         """
         copy = Dataset.copy(self, _deep_copy=True, _as=MaterializedDataset)
         copy._plan.execute(force_read=True)
+
+        blocks = copy._plan._snapshot_blocks
+
+        def get_input_data():
+            from ray.data._internal.execution.interfaces import RefBundle
+
+            if not blocks:
+                return []
+            else:
+                # TODO(hchen): Here we generate the same number of blocks as
+                # the original Dataset. Because the old code path does this, and
+                # some unit tests implicily depend on this behavior.
+                # After we remove the old code path, we should consider merging
+                # some blocks for better perf.
+                return [
+                    RefBundle(
+                        blocks=[block_with_metadata],
+                        owns_blocks=False,
+                    )
+                    for block_with_metadata in blocks.get_blocks_with_metadata()
+                ]
+
+        # Create a new logical plan whose input is the existing data
+        # from the the old Dataset.
+        copy._logical_plan = LogicalPlan(InputData(input_data_factory=get_input_data))
```

Comment:
Why not create the blocks right away during the materialize call?

Reply:
I thought `input_data_factory` would delay the RefBundle conversion to a remote worker. Just checked the code and confirmed it's not the case. So better create the RefBundles directly here.